src/viam/sdk/common/exception.hpp

```diff
@@ -0,0 +1,84 @@
+/// @file common/exception.hpp
+///
+/// @brief Defines custom exceptions for the SDK.
+#pragma once
+#include <stdexcept>
+#include <string>
+
+#include <viam/sdk/resource/resource_api.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Exception Classes related to SDK exceptions.
+
+/// @class ViamException
+/// @brief Defines a base exception type to be inherited from by other more
+/// specific exceptions.
+/// @ingroup Exception
+class ViamException : public std::runtime_error {
+   public:
+    explicit ViamException(const std::string& what);
+    explicit ViamException(const std::string& what, const std::string& type);
+    virtual ~ViamException();
+};
+
+/// @class ConnectionException
+/// @brief An exception thrown when an SDK client cannot create a connection to
+/// a machine.
+/// @ingroup Exception
+class ConnectionException : public ViamException {
+   public:
+    explicit ConnectionException(const std::string& what);
+};
+
+/// @class DuplicateRegistrationException
+/// @brief An exception thrown when an API/model pair has already been registered.
+/// @ingroup Exception
+class DuplicateRegistrationException : public ViamException {
+   public:
+    explicit DuplicateRegistrationException(const std::string& what);
+};
+
+/// @class DuplicateResourceException
+/// @brief An exception thrown when a resource has already been added to a
+/// resource manager.
+/// @ingroup Exception
+class DuplicateResourceException : public ViamException {
+   public:
+    explicit DuplicateResourceException(const std::string& what);
+};
+
+/// @class GRPCException
+/// @brief An exception thrown when a machine has sent an error back to an SDK
+/// client over gRPC.
+/// @ingroup Exception
+class GRPCException : public ViamException {
+   public:
+    explicit GRPCException(const std::string& what, int grpc_error_code);
+
+    int error_code() const noexcept;
```

Comment:
I'm not sure how important this method nor the `grpc_error_code_` field are. I can get rid of them if we see no use case for someone catching a `GRPCException` and accessing its error code.

Reply:
After switching to using `std::error_code`, I've lost the gRPC error code and only maintained the gRPC status message. I can re-add `grpc_error_code_` as an optional field on `ViamException` that's only present if it's a gRPC exception if desired?

Reply:
Hmm... I think it's a bit clunkier in the case where we don't have a `GRPCException` type specifically. Perhaps we could `string`ify the entire `grpc::Status` and pass that as the `what`? I want to think about it a bit but I'm a bit hesitant on the optional "only-relevant-if-grpc" field.